components/proxy/src/main/java/com/hotels/styx/admin/handlers/MetricsHandler.java

```diff
@@ -68,37 +70,35 @@ public MetricsHandler(CodaHaleMetricRegistry metricRegistry, Optional<Duration>
 
     @Override
     public Observable<HttpResponse> handle(HttpRequest request) {
-        return metricName(request.path())
-                .map(metricName -> specificMetricsResponse(request, metricName))
-                .orElseGet(() -> super.handle(request));
-    }
+        MetricRequest metricRequest = new MetricRequest(request);
 
-    private Observable<HttpResponse> specificMetricsResponse(HttpRequest request, String metricName) {
-        Map<String, Metric> metrics = metrics(metricName);
-
-        if (metrics.isEmpty()) {
-            return just(response(NOT_FOUND).build().toStreamingResponse());
-        }
+        return metricRequest.fullMetrics()
+                ? super.handle(request)
+                : just(restrictedMetricsResponse(metricRequest).build().toStreamingResponse());
+    }
 
-        boolean pretty = request.queryParam("pretty").isPresent();
+    private FullHttpResponse.Builder restrictedMetricsResponse(MetricRequest request) {
+        Map<String, Metric> fullMetrics = metricRegistry.getMetricRegistry().getMetrics();
 
-        String serialised = serialise(metrics, pretty);
+        Map<String, Metric> restricted = filter(fullMetrics, (name, metric) -> request.matchesRoot(name));
 
-        return just(response(OK)
-                .body(serialised, UTF_8)
-                .disableCaching()
-                .build()
-                .toStreamingResponse());
+        return restricted.isEmpty()
+                ? response(NOT_FOUND)
+                : search(request, restricted);
     }
 
-    private static Optional<String> metricName(String path) {
-        Matcher matcher = SPECIFIC_METRICS_PATH_PATTERN.matcher(path);
+    private FullHttpResponse.Builder search(MetricRequest request, Map<String, Metric> metrics) {
+        Map<String, Metric> searched = filter(metrics, (name, metric) -> request.containsSearchTerm(name));
 
-        if (matcher.matches()) {
-            return Optional.of(matcher.group(1));
-        }
+        String body = serialise(searched, request.prettyPrint);
 
-        return Optional.empty();
+        return response(OK)
+                .body(body, UTF_8)
+                .disableCaching();
+    }
+
+    private static <K, V> Map<K, V> filter(Map<K, V> map, BiPredicate<K, V> predicate) {
```

Comment:
We need to refactor this code.

Reply:
In what way?

Reply:
Well, this handler should perform three logical steps:

1. Query for the metrics
2. Filter results
3. Serialise and respond

But these logical steps disappear within the layers of abstraction. I'll find this class rather confusing considering its really simple responsibility.

Reply:
Anyway, I will approve. Let's merge and get back to this another time.